controlplane/kubeadm/internal/filters.go

```diff
@@ -41,65 +40,73 @@ import (
+		logReasons = append(logReasons, reason)
+		conditionReasons = append(conditionReasons, "spec.infrastructureRef is not up to date")
 	}
 
-	if len(mismatchReasons) > 0 {
-		return strings.Join(mismatchReasons, ","), false, nil
+	if len(logReasons) > 0 || len(conditionReasons) > 0 {
+		return false, logReasons, conditionReasons, nil
 	}
 
-	return "", true, nil
+	return true, nil, nil, nil
 }
 
 // NeedsRollout checks if a Machine needs to be rolled out and returns the reason why.
-func NeedsRollout(reconciliationTime, rolloutAfter *metav1.Time, rolloutBefore *controlplanev1.RolloutBefore, infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane, machine *clusterv1.Machine) (string, bool, error) {
-	rolloutReasons := []string{}
+func NeedsRollout(reconciliationTime, rolloutAfter *metav1.Time, rolloutBefore *controlplanev1.RolloutBefore, infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane, machine *clusterv1.Machine) (bool, []string, []string, error) {
+	logReasons := []string{}
+	conditionReasons := []string{}
 
 	// Machines whose certificates are about to expire.
 	if collections.ShouldRolloutBefore(reconciliationTime, rolloutBefore)(machine) {
-		rolloutReasons = append(rolloutReasons, "certificates will expire soon, rolloutBefore expired")
+		logReasons = append(logReasons, "certificates will expire soon, rolloutBefore expired")
+		conditionReasons = append(conditionReasons, "certificates will expire soon")
 	}
 
 	// Machines that are scheduled for rollout (KCP.Spec.RolloutAfter set,
 	// the RolloutAfter deadline is expired, and the machine was created before the deadline).
 	if collections.ShouldRolloutAfter(reconciliationTime, rolloutAfter)(machine) {
-		rolloutReasons = append(rolloutReasons, "rolloutAfter expired")
+		logReasons = append(logReasons, "rolloutAfter expired")
+		conditionReasons = append(conditionReasons, "kcp.spec.rolloutAfter expired")
 	}
 
 	// Machines that do not match with KCP config.
-	mismatchReason, matches, err := matchesMachineSpec(infraConfigs, machineConfigs, kcp, machine)
+	matches, specLogReasons, specConditionReasons, err := matchesMachineSpec(infraConfigs, machineConfigs, kcp, machine)
 	if err != nil {
-		return "", false, errors.Wrapf(err, "failed to determine if Machine %s needs rollout", machine.Name)
+		return false, nil, nil, errors.Wrapf(err, "failed to determine if Machine %s needs rollout", machine.Name)
 	}
 	if !matches {
-		rolloutReasons = append(rolloutReasons, mismatchReason)
+		logReasons = append(logReasons, specLogReasons...)
+		conditionReasons = append(conditionReasons, specConditionReasons...)
 	}
 
-	if len(rolloutReasons) > 0 {
-		return fmt.Sprintf("Machine %s needs rollout: %s", machine.Name, strings.Join(rolloutReasons, ",")), true, nil
+	if len(logReasons) > 0 {
```

Comment:
Above we check also for conditionReasons > 0 (but this would be unnecessary if we just have a struct for condition and log reasons)